woocommerce/woocommerce
# [Settings] Use DataForms to render settings fields

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Related to https://github.com/woocommerce/woocommerce/issues/52316

This PR shows how we can use DataForms to render settings fields instead of custom code. 

To use DataForm, we need to ensure each setting field has an `id` property, so I updated the setting data transformer to add a random id if it's missing.


https://github.com/user-attachments/assets/c466cbc0-0b44-47af-a3e2-a244303137b7


---

DataForm is a component that allows us to render form fields with a consistent UI. It supports a variety of field types, including text, number, checkbox, and select. Additionally, it's possible to render custom fields by providing a custom component. 

For more details, see:

- [DataForm README](https://github.com/WordPress/gutenberg/blob/227fcd87f1eccf4b853850c8f439fd95ba1e5f5f/packages/dataviews/README.md#dataform)
- [DataForm storybook](https://wordpress.github.io/gutenberg/?path=/docs/dataviews-dataform--docs&args=type:regular)

`@wordpress/dataviews@4.10.0` has a bug in package.json that causes the tests to fail. It has been fixed in https://github.com/WordPress/gutenberg/pull/67962 but not yet released so I've created a patch to fix it for now. Once it's released, we can remove the patch.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Add/Enable Gutenberg version `19.8.0` or earlier.
2. Turn on the `settings` feature flag
3. Use the [Settings Tester](https://github.com/woocommerce/woocommerce-settings-migration-tester) to test Modern screens. Clone it locally and `npm install && npm run plugin-zip` / [settings-tester.zip](https://github.com/user-attachments/files/18208344/settings-tester.zip)
3. Upload and activate the Tester
4. Go to `WooCommerce > Settings` and poke around and see checkboxes, selects, and inputs rendered with values.
5. Go to `WooCommerce > Settings > Emails` and see a custom HTML table rendered.
6. Next go to Site visibility to confirm our slotFill root is available by querying the document in the console.

```js
document.getElementById('wc_settings_site_visibility_slotfill');
```

7. Go to `/wp-admin/admin.php?page=wc-settings&tab=settings-tester&section=javascript`
8. Toggle `Echoing HTML` checkbox and confirm the yellow background is shown when the checkbox is checked.



### Changelog entry












<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 
Use DataForms to render settings fields 

</details>
